inginious/frontend/pages/course_admin/student_list.py

```diff
@@ -192,33 +198,48 @@ def post_audiences(self, course, data, active_tab, msg, error):
             active_tab = "tab_audiences"
 
         try:
-            if "upload_audiences" in data or "audiences" in data:
-                errored_students = []
-                if "upload_audiences" in data:
+            if "audiencefile" in data and 'upload_audiences_creation' in data:
+                # get the Werkzeug datastructures.FileStorage object.
+                # The stream of this object is the stream body of the uploaded file.
+                # Furthermore, FileStorage.stream seems to inherit ʻio.BufferedIOBase`, so this stream should be boiled.
+                reader = csv.reader(io.TextIOWrapper(data["audiencefile"], encoding='utf-8'))
+                csv_data = list(reader)  # read everything else into a list of rows
```

Comment:
This line is probably not needed as reader is a generator. If this is the case it also worsens performance to cast it into a list.

Reply:
reader is an iterator : https://docs.python.org/3/library/csv.html#csv.reader and as I have to iterate the data twice, it seems faster to cast into a list in place of using tee() for example (https://docs.python.org/3/library/itertools.html#itertools.tee).